### PR TITLE
Adds ExpressLRS Support

### PR DESCRIPTION
**Pull request by AlessandroAU**
First attempt at ELRS support for review/comments. Tested and working on bench. Should work with JP5in1 etc, although note that this was prototyped on a DIY board consisting of a iRangex4in1 + rfm95w module. 

Supports 200hz, 100hz, 50hz, no telemetry implemented yet. 
Binding also not implemented yet, will connected to a ExpressLRS RX using the default bind phrase. (Conventional binding not yet merged into ExpressLRS quite yet anyway).

Sadly, it overflows flash with all other protos enabled. (122172 bytes)

### CODE REVIEW

**Comment by howels** [comment]
Is this PR still in-progress? Recent posts regarding ExpressLRS have highlighted this project and I'm wondering if this might work on the Jumper T18.

**Comment by AlessandroAU** [comment]
Yeah, main problem is that I don't have a T18 for testing and this currently overflows the flash size, so other protos would have to be removed to fit.

**Comment by benlye** [comment]
> Yeah, main problem is that I don't have a T18 for testing and this currently overflows the flash size, so other protos would have to be removed to fit.

It won't be long and any new protocol will overflow the flash on all STM32 modules, but the 5in1 is particularly tight. We've bought some time by eliminating USB support, but it's a temporary stay at best. 

So don't worry about that happening, because it's inevitable anyway :-)

**Comment by gretel** [comment]
@AlessandroAU me has as `uruav mpm` with the [wrong mcu](https://www.multi-module.org/basics/module-hardware/unsupported-modules) 😝 therefore i do compile binaries of this which are just a few kilobytes.. will not fix the emulated eeprom, though.
long story short i have a module to mess with cause it can't get worse - looking forward to test.
btw, did you get any chance to check on possible compatbility of the `tbs mpm` module just released?

File: Multiprotocol/ExpressLRS.ino [hunk 1]
```diff
@@ -0,0 +1,540 @@
+#if defined(ELRS_SX1276_INO) && defined(SX1276_INSTALLED)
+
+#include "iface_sx1276.h"
+#include "ExpressLRS.h"
+
+uint8_t UID[6] = {10, 183, 12, 124, 56, 3}; // default UID for ExpressLRS (for testing)
```
Comment (gretel) [inline Multiprotocol/ExpressLRS.ino:6]: this is just `default`, not `default ExpressLRS binding phrase` as in https://github.com/AlessandroAU/ExpressLRS/blob/dc00b8ca9db7e67516de3469f2528d08ce36ea98/src/user_defines.txt#L13

made a script to get the `uid`:

```
binding phrase: 'default ExpressLRS binding phrase'
     uid bytes: '165,221,171,247,138,255'

Multiprotocol/ExpressLRS.ino:
  uint8_t UID[6] = {165,221,171,247,138,255};
```

```python
#!/usr/bin/env python3

import argparse
import hashlib
import sys

parser = argparse.ArgumentParser(prog='PROG')
parser.add_argument('--phrase', type=str, default='default ExpressLRS binding phrase', help='binding phrase')
args = parser.parse_args()

bindingPhraseHash = hashlib.md5(args.phrase.encode()).digest()
UIDbytes = (str(bindingPhraseHash[0]) + "," + str(bindingPhraseHash[1]) + "," + str(bindingPhraseHash[2]) + ","+ str(bindingPhraseHash[3]) + "," + str(bindingPhraseHash[4]) + "," + str(bindingPhraseHash[5]))

sys.stdout.write("binding phrase: '%s'\n     uid bytes: '%s'\n\n" % ( args.phrase, UIDbytes))
sys.stdout.write("Multiprotocol/ExpressLRS.ino:\n  uint8_t UID[6] = {%s};\n" % UIDbytes)
sys.stdout.flush()
```

**Comment by OpenUAS** [comment]
Well IMHO rather have ExpressLRS on the cost of some crappy CN obscure protocols removal, have some hardware to donate for testing

**Comment by AlessandroAU** [comment]
Thanks for all the input guys, I realise this PR is a bit out of date now. I will try to spend some time on this.

File: Multiprotocol/ExpressLRS.ino (hunk 1 above)
Comment (AlessandroAU) [reply Multiprotocol/ExpressLRS.ino:6]: Thanks dude! will update UID. Need to sort out eeprom then static UID can be done away with.

**Comment by gretel** [comment]
> Thanks for all the input guys, I realise this PR is a bit out of date now. I will try to spend some time on this.

cool :)

**Comment by chaotixtech** [comment]
The basic code from this pr was added into current, as of yesterday, MPM code by me and compiled and flashed with a lot of obscure protocols commented, onto an actual T18 radio. Bench test with a flashed r9mx proved bind and failsafe work. Please continue with the pr.

**Comment by harry1453** [comment]
@pascallanger other than obvious things like merge conflicts, what needs to be done for this to be acceptable?

**Comment by pascallanger** [comment]
@harry1453 this is not a question of acceptable or not. Alessandro told me a long time ago that the PR was not ready and still needed to be modified. So I'll basically waiting. I can merge it tomorrow if you say the code is working. I haven't really checked how things have been done and I'm fine to merge if people are happy with it.

**Comment by gretel** [comment]
> Thanks for all the input guys, I realise this PR is a bit out of date now. I will try to spend some time on this.

howdy @AlessandroAU - this PR got kinda stuck. can you put some time on this? i'd be super for the hobby. thanks a mill!

**Comment by joevo2** [comment]
Would be super cool if it works with 2.4GHz ExpressLRS. Btw does anyone is making off the shelf ExpressLRS receiver? If MPM supports this and we can get the receiver easier it'll be super awesome.

**Comment by hazcoper** [comment]
Anyone know what is the current status for using it with the 5 in 1 module with 2.4 or 900?

**Comment by jimsynz** [comment]
> Would be super cool if it works with 2.4GHz ExpressLRS. Btw does anyone is making off the shelf ExpressLRS receiver? If MPM supports this and we can get the receiver easier it'll be super awesome.

There's a bunch made by Happy Model on Banggood.  I am super hopeful that I'll be able to add ExpressLRS support to my RadioMaster TX at some point soon.

**Comment by Talha909** [comment]
when will we get LRS support in 4 in 1 Multi protocol

**Comment by harry1453** [comment]
@hazcoper The current status with 5in1 is that this PR allegedly worked at some point, at some point I'm going to get around to reviving this. You're welcome to give it a go if you know how.

@jimsynz As the radiomaster doesn't have an SX1276 or SX1280 chip onboard, the only way to use ELRS on it without modifying/replacing the internal module is to use an external module

@Talha909 for the 4in1, there will never be support, as the 4in1 does not include a SX1276 (included in the 5in1) or a SX1280 (not included in any production MPMs)

**Comment by iKonrad** [comment]
Hi, has there been any progress with the PR?

**Comment by harry1453** [comment]
@iKonrad no. I now have enough hardware to possibly start working on this, I'm just waiting on a) a stable v1.0 ELRS release and b) time to do it

**Comment by yokoboko** [comment]
Any news on this one?

**Comment by jasperchua99** [comment]
Any updates?

**Comment by AlessandroAU** [comment]
No one is currently working on this

**Comment by gnuton** [comment]
I have updated the code of this PR and now it could be merged with current upstream  master
Please find my changes here:
https://github.com/gnuton/DIY-Multiprotocol-TX-Module/tree/jumper-5-in-1-elrs

NOTE: At moment I do not have a jumper t18s in my hands, so the code in my branch even if builds fine it has not been tested.

**Comment by shafr** [comment]
@gnuton what help do you need to finish this one ? 
How do I select integrated ELRS protocol to test ?

I can build FW & run test on my Jumper t-pro (if it's supported)

**Comment by gnuton** [comment]
@shafr I have not tested this on the J T-Pro yet. but you are brave enough you can build,flash and you should see ELRS among the protocols in the multiprotocol list.
Currently this code needs a major rewrite sine it would work only with an old version of ELRS from June 2020 IIRC..which means you have to flash a RX with such old ELRS firmware to see something working.

**Comment by shafr** [comment]
@gnuton actually I do have an:
* Jumper t-pro with latest EdgeTx 2.80.
* Cetus X drone with ELRS 2.* something version.

Can you please provide with compiled firmware to upload ? I was hoping there would be some github action or docker image that I can use to build instead of doing all the manual steps.

**Comment by pfeerick** [comment]
You'd need something that can run a much earlier version of ELRS - June 2020 puts this before v1.0 of ELRS (Jul 17, 2021). However, the reality is due to the MPM architecture it won't be possible to fully support all of the ELRS features, so there will be little incentive when either using a dedicated module or replacing internal RF with an ELRS module results in better end-user experience.

**Comment by shafr** [comment]
OK, understood, so effort of implementing this is not worth it, everyone can just buy another controller / external module.
Maybe this can be added to README or pinned to issues stack so this would not be asked every time.
Thank you!

**Comment by OpenUAS** [comment]
Close the issue maybe?

**Comment by pascallanger** [comment]
Not adding this protocol since ELRS have their own module.

**Pull request closed without merging**